lib/smack/SplitAggregateValue.cpp

```diff
@@ -0,0 +1,139 @@
+  for(Instruction& I : BB) {
+    IndexT idx;
+    std::vector<InfoT> info;
+    if (LoadInst* li = dyn_cast<LoadInst>(&I)) {
+      if (li->getType()->isAggregateType()) {
+        visitAggregateValue(nullptr, li->getType(), idx, info, C);
+        splitAggregateLoad(li, info);
+        toRemove.push_back(li);
+      }
+    } else if (StoreInst* si = dyn_cast<StoreInst>(&I)) {
+      Value* V = si->getValueOperand();
+      if (V->getType()->isAggregateType()) {
+        visitAggregateValue(dyn_cast_or_null<Constant>(V), V->getType(), idx, info, C);
+        splitAggregateStore(si, info);
+        toRemove.push_back(si);
+      }
+    } else if (ReturnInst* ri = dyn_cast<ReturnInst>(&I)) {
+      Value* V = ri->getReturnValue();
+      if (auto crv = dyn_cast_or_null<ConstantAggregate>(V)) {
+        visitAggregateValue(cast<Constant>(V), V->getType(), idx, info, C);
+        splitConstantReturn(ri, info);
+      }
+    }
+  }
+
+  for (auto& i : toRemove)
+    i->eraseFromParent();
+  return true;
+}
+
+void SplitAggregateValue::splitAggregateLoad(LoadInst* li, std::vector<InfoT>& info) {
+  IRBuilder<> irb(li);
+  Value* V = UndefValue::get(li->getType());
+  Value* P = li->getPointerOperand();
+  for (auto& e : info) {
+    IndexT idxs = std::get<0>(e);
+    V = irb.CreateInsertValue(V,
+      irb.CreateLoad(irb.CreateGEP(P, ArrayRef<Value*>(getFirsts(idxs)))),
+      ArrayRef<unsigned>(getSeconds(idxs)));
+  }
+  li->replaceAllUsesWith(V);
+}
+
+void SplitAggregateValue::splitAggregateStore(StoreInst* si, std::vector<InfoT>& info) {
+  IRBuilder<> irb(si);
+  Value* P = si->getPointerOperand();
+  Value* V = si->getValueOperand();
+  for (auto& e : info) {
+    IndexT idxs = std::get<0>(e);
+    Constant* c = std::get<1>(e);
+    std::vector<Value*> vidxs = getFirsts(idxs);
+    if (c)
+      irb.CreateStore(c, irb.CreateGEP(P, ArrayRef<Value*>(vidxs)));
+    else
+      irb.CreateStore(irb.CreateExtractValue(
+        V, ArrayRef<unsigned>(getSeconds(idxs))),
+          irb.CreateGEP(P, ArrayRef<Value*>(vidxs)));
+  }
+}
+
+void SplitAggregateValue::splitConstantReturn(ReturnInst* ri, std::vector<InfoT>& info) {
+  IRBuilder<> irb(ri);
+  Type* T = ri->getReturnValue()->getType();
+  Value* V = UndefValue::get(T);
+  Value* box = irb.CreateAlloca(T);
+  for (auto& e : info) {
+    IndexT idxs = std::get<0>(e);
+    Constant* c = std::get<1>(e);
+    Value* P = irb.CreateGEP(box, ArrayRef<Value*>(getFirsts(idxs)));
+    irb.CreateStore(c, P);
+    V = irb.CreateInsertValue(V, irb.CreateLoad(P),
+      ArrayRef<unsigned>(getSeconds(idxs)));
+  }
+  ri->setOperand(0, V);
+}
+
+void SplitAggregateValue::visitAggregateValue(Constant* baseVal, Type* T, IndexT idxs,
+                                                    std::vector<InfoT>& info, LLVMContext& C){
+  Constant* newBaseVal = baseVal;
+  if(T->isIntegerTy() || T->isFloatingPointTy() || T->isPointerTy())
+      info.push_back(std::make_pair(idxs, newBaseVal));
```

Comment:
While you’re here, why not improve this code using uniform initialization, e.g., replacing every instance of `std::make_pair(A,B)` with `{A, B}`.